src/Compilers/CSharp/Portable/FlowAnalysis/DefinitelyAssignedOnEntryWalker.cs

```diff
@@ -0,0 +1,99 @@
+    internal class DefinitelyAssignedOnEntryWalker : AbstractRegionDataFlowPass
+    {
+        private readonly HashSet<Symbol> _definitelyAssignedOnEntry = new HashSet<Symbol>();
+
+        private DefinitelyAssignedOnEntryWalker(
+            CSharpCompilation compilation,
+            Symbol member,
+            BoundNode node,
+            BoundNode firstInRegion,
+            BoundNode lastInRegion)
+            : base(compilation, member, node, firstInRegion, lastInRegion)
+        {
+        }
+
+        internal static HashSet<Symbol> Analyze(
+            CSharpCompilation compilation, Symbol member, BoundNode node, BoundNode firstInRegion, BoundNode lastInRegion, out bool? succeeded)
+        {
+            var walker = new DefinitelyAssignedOnEntryWalker(compilation, member, node, firstInRegion, lastInRegion);
+            try
+            {
+                bool badRegion = false;
+                var result = walker.Analyze(ref badRegion);
+                succeeded = !badRegion;
+                return badRegion ? new HashSet<Symbol>() : result;
+            }
+            finally
+            {
+                walker.Free();
+            }
+        }
+
+        private HashSet<Symbol> Analyze(ref bool badRegion)
+        {
+            base.Analyze(ref badRegion, null);
+            return _definitelyAssignedOnEntry;
+        }
+
+        protected override void EnterRegion()
+        {
+            // this can happen multiple times as flow analysis is multi-pass.  Always 
+            // take the latest data and use that to determine our result.
+            _definitelyAssignedOnEntry.Clear();
+
+            if (this.IsConditionalState)
+            {
+                // We're in a state where there are different flow paths (i.e. when-true and when-false).
+                // In that case, a variable is only definitely assigned if it's definitely assigned through
+                // both paths.
+                this.ProcessState(this.StateWhenTrue, this.StateWhenFalse);
+            }
+            else
+            {
+                this.ProcessState(this.State, state2opt: null);
+            }
+
+            base.EnterRegion();
+        }
+
+        private void ProcessState(LocalState state1, LocalState state2opt)
+        {
+            foreach (var slot in state1.Assigned.TrueBits())
+            {
+                if (slot >= variableBySlot.Length)
+                {
+                    continue;
+                }
+
+                // It's not definitely assigned if it's not assigned through both 
+                // when-true and when-false paths.
+                if (state2opt != null && !state2opt.IsAssigned(slot))
+                {
+                    continue;
+                }
+
+                var variable = variableBySlot[slot];
+                if (variable.Exists)
+                {
+                    var symbol = variable.Symbol;
+                    if (symbol.Kind != SymbolKind.Field)
```

Comment:
I copied this bit from analysis passes.  i presume it makes sense here as well and we only track htis assignment information for non-fields.

Reply:
we track fields of structs, but don't want to report them in the API.

---
In reply to: [305634900](https://github.com/dotnet/roslyn/pull/37395#discussion_r305634900) [](ancestors = 305634900)